File: ios/Woona/Protocol/PacketValidator.swift
import Foundation

struct ValidatedPacket: Equatable {
    let bytes: [UInt8]
    let counter: UInt64
    let timerMillis: UInt64
    let measurementCount: Int
    let sensorBlocks: [SensorBlock]

    func channelSamples(sensorType: Int, channel: Int) -> [Float] {
        guard channel > 0 else { return [] }

        return sensorBlocks.flatMap { block -> [Float] in
            guard block.sensorType == sensorType, channel <= block.channelSamples.count else {
                return []
            }
            return block.channelSamples[channel - 1]
        }
    }
}

enum PacketValidationFailureReason: String, Equatable {
    case invalidStart = "Invalid packet start marker"
    case invalidLength = "Invalid packet length"
    case invalidMeasurementCount = "Invalid measurement count"
    case invalidSensorBlocks = "Invalid sensor block layout"
}

enum PacketValidationResult: Equatable {
    case accepted(ValidatedPacket)
    case rejected(PacketValidationFailureReason)
}

struct PacketValidator {
    private let sensorBlockParser: SensorBlockParser

    init(sensorBlockParser: SensorBlockParser = SensorBlockParser()) {
        self.sensorBlockParser = sensorBlockParser
    }

    func validate(_ packetBytes: [UInt8]) -> PacketValidationResult {
        validate(packetBytes, parseResult: sensorBlockParser.parseResult(packetBytes))
    }

    func validate(_ packetBytes: [UInt8], sensorBlocks: [SensorBlock]) -> PacketValidationResult {
        validate(packetBytes, parseResult: sensorBlockParser.parseResult(packetBytes), parsedSensorBlocks: sensorBlocks)
    }

    private func validate(
        _ packetBytes: [UInt8],
        parseResult: SensorBlockParseResult,
        parsedSensorBlocks: [SensorBlock]? = nil
    ) -> PacketValidationResult {
        guard packetBytes.count >= Self.minimumPacketSize else {
            return .rejected(.invalidLength)
        }

        guard Self.hasStartMarker(packetBytes, at: 0) else {
            return .rejected(.invalidStart)
        }

        let declaredLength = Self.decodeLength(packetBytes, at: 0)
        guard declaredLength == packetBytes.count,
              (Self.minimumPacketSize...Self.maximumPacketSize).contains(declaredLength) else {
            return .rejected(.invalidLength)
        }

        let measurementCount = Int(packetBytes[Self.measurementCountOffset])
        guard (1...Self.maximumMeasurementCount).contains(measurementCount) else {
            return .rejected(.invalidMeasurementCount)
        }

        guard isValidSensorBlockLayout(parseResult, measurementCount: measurementCount, packetLength: packetBytes.count) else {
            return .rejected(.invalidSensorBlocks)
        }

        return .accepted(
            ValidatedPacket(
                bytes: packetBytes,
                counter: Self.decodeCounter(packetBytes, at: 0),
                timerMillis: Self.decodeTimer(packetBytes, at: 0),
                measurementCount: measurementCount,
                sensorBlocks: parsedSensorBlocks ?? parseResult.blocks
            )
        )
    }

    private func isValidSensorBlockLayout(
        _ parseResult: SensorBlockParseResult,
        measurementCount: Int,
        packetLength: Int
    ) -> Bool {
        guard parseResult.parsedBlockCount == measurementCount else { return false }
        guard parseResult.blocks.count == measurementCount else { return false }
        guard parseResult.blocks.allSatisfy({ !$0.channelSamples.isEmpty }) else { return false }

        let trailerByteCount = packetLength - parseResult.consumedBytes
        return (0...Self.maximumSensorBlockTrailerBytes).contains(trailerByteCount)
    }

    static func hasStartMarker(_ bytes: [UInt8], at offset: Int) -> Bool {
        guard offset >= 0, offset + startMarker.count <= bytes.count else { return false }
        return bytes[offset..<(offset + startMarker.count)].elementsEqual(startMarker)
    }

    static func decodeLength(_ bytes: [UInt8], at offset: Int = 0) -> Int {
        Int(bytes[offset + lengthOffset]) | (Int(bytes[offset + lengthOffset + 1]) << 8)
    }

    static func decodeCounter(_ bytes: [UInt8], at offset: Int = 0) -> UInt64 {
        UInt64(bytes[offset + counterOffset]) |
            (UInt64(bytes[offset + counterOffset + 1]) << 8) |
            (UInt64(bytes[offset + counterOffset + 2]) << 16) |
            (UInt64(bytes[offset + counterOffset + 3]) << 24)
    }

    static func decodeTimer(_ bytes: [UInt8], at offset: Int = 0) -> UInt64 {
        UInt64(bytes[offset + timerOffset]) |
            (UInt64(bytes[offset + timerOffset + 1]) << 8) |
            (UInt64(bytes[offset + timerOffset + 2]) << 16) |
            (UInt64(bytes[offset + timerOffset + 3]) << 24)
    }

    static let startMarker: [UInt8] = [0x33, 0x99, 0xAA, 0x55]
    static let minimumPacketSize = 16
    static let maximumPacketSize = 16_384
    static let maximumMeasurementCount = 4
    static let maximumSensorBlockTrailerBytes = 4
    static let lengthOffset = 4
    static let measurementCountOffset = 6
    static let counterOffset = 7
    static let timerOffset = 11
}

File: android/app/src/main/java/com/example/myapplication/protocol/PacketValidator.kt
package com.example.myapplication.protocol

data class ValidatedPacket(
    val bytes: ByteArray,
    val counter: Long,
    val timerMillis: Long,
    val measurementCount: Int,
    val sensorBlocks: List<SensorBlock>,
) {
    fun channelSamples(
        sensorType: Int,
        channel: Int,
    ): List<Float> {
        if (channel <= 0) return emptyList()

        return sensorBlocks
            .asSequence()
            .filter { it.sensorType == sensorType && channel <= it.channelSamples.size }
            .flatMap { it.channelSamples[channel - 1].asSequence() }
            .toList()
    }
}

enum class PacketValidationFailureReason(val description: String) {
    INVALID_START("Invalid packet start marker"),
    INVALID_LENGTH("Invalid packet length"),
    INVALID_MEASUREMENT_COUNT("Invalid measurement count"),
}

sealed interface PacketValidationResult {
    data class Accepted(
        val packet: ValidatedPacket,
    ) : PacketValidationResult

    data class Rejected(
        val reason: PacketValidationFailureReason,
    ) : PacketValidationResult
}

class PacketValidator(
    private val sensorBlockParser: SensorBlockParser = SensorBlockParser(),
) {
    fun validate(packetBytes: ByteArray): PacketValidationResult {
        if (packetBytes.size < MIN_PACKET_SIZE) {
            return PacketValidationResult.Rejected(PacketValidationFailureReason.INVALID_LENGTH)
        }

        if (!hasStartMarker(packetBytes)) {
            return PacketValidationResult.Rejected(PacketValidationFailureReason.INVALID_START)
        }

        val declaredLength = decodeLength(packetBytes)
        if (declaredLength != packetBytes.size) {
            return PacketValidationResult.Rejected(PacketValidationFailureReason.INVALID_LENGTH)
        }

        val measurementCount = packetBytes[MEASUREMENT_COUNT_OFFSET].toInt() and 0xFF
        if (measurementCount !in 1..MAX_MEASUREMENT_COUNT) {
            return PacketValidationResult.Rejected(PacketValidationFailureReason.INVALID_MEASUREMENT_COUNT)
        }

        return PacketValidationResult.Accepted(
            packet = ValidatedPacket(
                bytes = packetBytes,
                counter = decodeCounter(packetBytes),
                timerMillis = decodeTimer(packetBytes),
                measurementCount = measurementCount,
                sensorBlocks = sensorBlockParser.parse(packetBytes),
            ),
        )
    }

    private fun hasStartMarker(packetBytes: ByteArray): Boolean {
        return packetBytes[0] == 0x33.toByte() &&
            packetBytes[1] == 0x99.toByte() &&
            packetBytes[2] == 0xAA.toByte() &&
            packetBytes[3] == 0x55.toByte()
    }

    private fun decodeLength(packetBytes: ByteArray): Int {
        return (packetBytes[LENGTH_OFFSET].toInt() and 0xFF) or
            ((packetBytes[LENGTH_OFFSET + 1].toInt() and 0xFF) shl 8)
    }

    private fun decodeCounter(packetBytes: ByteArray): Long {
        val c0 = packetBytes[COUNTER_OFFSET].toLong() and 0xFF
        val c1 = packetBytes[COUNTER_OFFSET + 1].toLong() and 0xFF
        val c2 = packetBytes[COUNTER_OFFSET + 2].toLong() and 0xFF
        val c3 = packetBytes[COUNTER_OFFSET + 3].toLong() and 0xFF
        return c0 or (c1 shl 8) or (c2 shl 16) or (c3 shl 24)
    }

    private fun decodeTimer(packetBytes: ByteArray): Long {
        val t0 = packetBytes[TIMER_OFFSET].toLong() and 0xFF
        val t1 = packetBytes[TIMER_OFFSET + 1].toLong() and 0xFF
        val t2 = packetBytes[TIMER_OFFSET + 2].toLong() and 0xFF
        val t3 = packetBytes[TIMER_OFFSET + 3].toLong() and 0xFF
        return t0 or (t1 shl 8) or (t2 shl 16) or (t3 shl 24)
    }

    private companion object {
        const val MIN_PACKET_SIZE = 16
        const val MAX_MEASUREMENT_COUNT = 4
        const val LENGTH_OFFSET = 4
        const val MEASUREMENT_COUNT_OFFSET = 6
        const val COUNTER_OFFSET = 7
        const val TIMER_OFFSET = 11
    }
}
